tests/test-grammar-integration.cpp

```diff
@@ -44,21 +44,24 @@ static bool test_build_grammar_fails(const std::string & grammar_str) {
     return grammar_fails;
 }
 
-static bool match_string(const std::string & input, llama_grammar* grammar) {
+static bool match_string(const std::string & input, llama_grammar * grammar) {
     auto decoded = decode_utf8(input, {});
 
     const auto & code_points = decoded.first;
 
+    llama_grammar_stacks & cur_stacks = llama_grammar_get_stacks(grammar);
+
```

Comment:
Not a significant optimization, but would it work to move the fetching of the rules to outside of the loop (alongside `cur_stacks`), since I don't think the rules will change inside of the loop?

```suggestion
    llama_grammar_stacks & cur_stacks = llama_grammar_get_stacks(grammar);
    const llama_grammar_rules  & prev_rules  = llama_grammar_get_rules (grammar);
```